pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -475,14 +482,35 @@ func (nc NodeController) tryUpdateNodeStatus(node *api.Node) (error, time.Durati
 			LastTransitionTime: node.CreationTimestamp,
 		}
 		gracePeriod = nodeStartupGracePeriod
+		nc.nodeStatusMap[node.Name] = NodeStatusData{
+			status:    node.Status,
+			timestamp: node.CreationTimestamp,
+		}
 	} else {
 		// If ready condition is not nil, make a copy of it, since we may modify it in place later.
 		lastReadyCondition = *readyCondition
 		gracePeriod = nodeMonitorGracePeriod
 	}
 
-	// Check last time when NodeReady was updated.
-	if nc.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) {
+	oldNodeStatus, found := nc.nodeStatusMap[node.Name]
+	if !found {
+		glog.Errorf("NodeController state corrupted: timestamp map misses entry for %s. Assuming now.", node.Name)
+		oldNodeStatus = NodeStatusData{
+			status:    node.Status,
+			timestamp: nc.now(),
+		}
+		nc.nodeStatusMap[node.Name] = oldNodeStatus
+	} else if !api.Semantic.DeepEqual(oldNodeStatus.status, node.Status) {
```

Comment:
IIUC this doesn't distinguish between a change made to the status by NodeController vs. a change made by Kubelet. I think that may cause a problem. For example consider the following scenario 

1) Kubelet writes NodeReady = true to etcd (via API server)
2) This function runs, and sets nc.nodeStatusMap[node.Name] = { stuff including NodeReady = true }
3) Machine fails
3) NodeController writes NodeReady = Unknown to etcd
4) This function runs. IIUC, oldNodeStatus.status == { nodeReady = true } and node.Status == { nodeReady = Unkonwn }. Since they are not equal, you will update NodeStatusData{timestamp} to nc.now(), which is wrong.

Assuming my analysis is correct, then I think one solution is to change this line to
```
} else if !api.Semantic.DeepEqal(...) && !node.Status.NodeReady == Unknown
```
(obviously that's pseudocode, but you get the idea)
I think this works because of the following facts
i) The only change to NodeStatus that NodeController makes is to set NodeReady = Unknown
ii) Kubelet never sets NodeReady = Unknown

Actually, thinking some more, I think a better solution would be to just diff oldNodeStatus.timestamp against the LastProbeTime in the NodeReady condition of the node.Status (rather than diffing the whole NodeStatus object). I think this has the same effect but is more intuitive because it makes it clear that nodeStatusMap.timestamp is storing is the local NodeController timestamp corresponding to the LastProbeTime in NodeReady.